### DISCUSSION TITLE
Version 3 Planning - Feature Requests and Language Design Planning

### DISCUSSION THREAD

**Discussion by lyxal** [Vyxal/Vyxal#551] [Version 3]
This is for planning what Vyxal 3.x will look like. Unlike usual feature request issues, you have the chance to completely reshape the language here (within reason).

Some things to note:

- v3 is still stack-based. That's a core feature
- You don't have to restrict yourself to the current codepage - that's up for reshaping too
- There's still going to be the existing structures (like if-statements, lambdas, etc). Those are also core features
- Type overloading is also still going to be a thing.

Other than that, everything is open to consideration. Think outside the box. Come up with the wildest/unique ideas you can.

I'd ask that you keep it to one suggestion per reply

**Comment by lyxal** [comment]
# Transformers

These are going to replace the current system of modifiers. The only difference is that transformers will be postfix like Jelly's quicks instead of prefix.

**Reply by cgccuser** [reply]
Wait, why make them postfix? Does it ease parsing or golfing?

**Reply by lyxal** [reply]
It greatly eases parsing - it means you can just pop elements off a stack rather than having to parse the rest of the program and then collect.

Also, it gives a more "tacit" feel - Jelly, APL, J, K etc all use postfix.

**Reply by pxeger** [reply]
Does this mean modifiers will be groupable like with Jelly's "last N links as a dyad" (although with no specificity on arity, presumably)?

**Reply by pxeger** [reply]
Will things like map, filter, etc. use this system too? What about loops, if statements, function definitions? (My vote is yes)

**Reply by lyxal** [reply]
> Does this mean modifiers will be groupable like with Jelly's "last N links as a dyad" (although with no specificity on arity, presumably)?

Yes. They kind of already are in 2.6+

**Reply by lyxal** [reply]
> Will things like map, filter, etc. use this system too? What about loops, if statements, function definitions? (My vote is yes)

Yes to this as well. They also kind of already do this in 2.6+ as well

**Reply by pxeger** [reply]
So can we free up any of `|;)]}` for generic elements?

**Reply by lyxal** [reply]
There'd still need to be one closing character just in case you wanted to use a structure without a transformer

**Reply by pxeger** [reply]
I thought you meant that structures would be replaced by transformers, so you'd have `ax[` instead of `a[x]`, and  `axy$[` (using Jelly `$`) instead of `a[xy]`. How would you use a structure without a transformer? Or did I misunderstand you?

**Reply by lyxal** [reply]
I meant as in `a[bxy]$` would be considered one unit

**Reply by pxeger** [reply]
Oh I see: so you can do like `[abcde]` instead of `a$b$c$d$e$]` or `abcdƲe$]`?

**Reply by lyxal** [reply]
correct. That's one of the core features of Vyxal anyway.

**Reply by chunkybanana** [reply]
A deep vectorise quick (vectorise all the way down).

**Reply by cgccuser** [reply]
Wait so wouldn't this be less golfy because you wouldn't be able to take advantage of auto-closing strings/structures?

**Reply by cgccuser** [reply]
(also minor note but prefix is actually easier to parse the way the Scala implementation of Vyxal is done right now lol (granted, it's not safe from stack overflows with really really huge programs))

**Reply by Steffan153** [reply]
Will we have one main closing character or will we stick with all of `;)]}`?

**Comment by lyxal** [comment]
# General Element Requests

Reply to this with any general element requests you'd like. They don't have to be specific with type overloads, just general ideas/summaries of what you think would make Vyxal golfier.

**Reply by mathcat4** [reply]
Is Vyxal going to support imaginary numbers?

**Reply by cgccuser** [reply]
I thought we already did?

**Reply by pxeger** [reply]
Not very well...

**Reply by lyxal** [reply]
@pxeger what do you mean? sympy's imaginary number system is used, and `°` can be used to create complex numbers

**Reply by lyxal** [reply]
[Try it Online!](https://vyxal.pythonanywhere.com/#WyIiLCIiLCIxwrA0IDXCsDIgKyIsIiIsIiJd)

**Reply by pxeger** [reply]
There's still a lot of optimising to do in terms of usability. Factorial of a complex number should use the gamma function. What happens when you compare complex numbers? I dunno.

**Reply by TheSecondComing123** [reply]
What about an element that transforms a string to match case with another? For example `"abc""XyZ"x` => `AbC`.

**Comment by chunkybanana** [comment]
# Better English compression

While our string compression is fine for stuff with words and other symbols, it's substantially worse than say Jelly's for plain english.

**Reply by Seggan** [reply]
I have managed to copy Jelly's compression for a now-obsolete language I was writing, we can use that. As long as the license allows it, of course.

**Reply by chunkybanana** [reply]
> As long as the license allows it, of course.

Jelly's MIT licensed, Vyxal's MIT licensed. No problemo.

**Comment by pxeger** [comment]
Here are a few patterns which can't easily be shrunk with new additions the way Vyxal currently works. I think these could be dealt with using the transformer systems:

- `{›...|`

**Comment by cgccuser** [comment]
Having lambdas operate on the global stack and functions on their own local stacks would be really nice. You wouldn't have to keep track of the arity of every lambda and element (you'd still to have keep track of function arities but that's much more manageable). Lambdas are also written right where they're applied, meaning that you're aware of what the global stack looks like at that point and can work right there. Functions are written elsewhere, so you have no idea what the global stack will be like when you run a function. Of course, we could always have a flag to make functions work on the global stack,  but not to make lambdas work on local stacks.

**Reply by lyxal** [reply]
How about a global lambda structure instead? Local stack lambdas allow for argument reuse to happen and means that you can remove instances of `n` from your lambda.

**Reply by chunkybanana** [reply]
*NO*. Lambdas are functions, they should stay that way .It's much more sensible for argument reuse anyway.

**Comment by sporeball** [comment]
## digraph to output a string, colored
such a digraph might take `str a, num b` as arguments, and return `\x1b[38;5;${NUM B}m${STR A}\x1b[0m`, i.e. `str a` colored using the [ANSI escape codes](https://en.wikipedia.org/wiki/ANSI_escape_code#8-bit) corresponding to the color mapped to `num b`

**Reply by zoomlogo** [reply]
This won't make sense on online interpreters though

**Reply by sporeball** [reply]
[naz's online interpreter](https://sporeball.dev/naz) uses `span`s with specific colors to make it feel more like the Node interpreter would; maybe Vyxal's online interpreter could do the same for the purposes of this digraph?\
it would be annoying to have to keep a list of 256 color codes somewhere, though...

**Reply by pxeger** [reply]
There are libraries to automatically convert ANSI coloured text to HTML markup. Even full terminal support could be emulated using xterm.js

**Reply by lyxal** [reply]
Also, there's an `online` attribute of `ctx` that determines if the online interpreter is being used, so it could switch between spans and ANSI.

**Comment by chunkybanana** [comment]
Probably just `;` for golfiness.

**Reply by Steffan153** [reply]
Based on the current YAML, it's `}`.

**Comment by PlaceReporter99** [comment]
# Slight lambda improvement
Maybe a regular lambda could apply to the next four elements if no closing semicolon is found for the lambda?
